libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/TripServiceImpl.kt

```diff
@@ -0,0 +1,145 @@
+internal object NavigationNotificationService : Service() {
+    private val job = SupervisorJob()
+    val serviceScope = CoroutineScope(job + Dispatchers.IO)
+
+    var locationEngineCallback = object : LocationEngineCallback<LocationEngineResult> {
+        override fun onSuccess(result: LocationEngineResult?) {
+        }
+
+        override fun onFailure(exception: java.lang.Exception) {
+        }
+    }
+    var locationEngine: LocationEngine? = null
+    val locationEngineRequest: LocationEngineRequest = LocationEngineRequest.Builder(UPDATE_INTERVAL_IN_MILLISECONDS)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .setFastestInterval(FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS)
+            .build()
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return null
+    }
+
+    override fun onCreate() {
+        startForeground(NAVIGATION_SERVICE_ID, null)
+        notificationChannel.close()
+        notificationChannel = ConflatedBroadcastChannel()
+        serviceScope.launch {
+            while (isActive) {
+                val status = MapboxNativeNavigatorImpl.getStatus(java.util.Date()) // Request that will drive the Polling functionality
+                locationEngine?.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {
+                    override fun onSuccess(result: LocationEngineResult?) {
+                        ifNonNull(result, result?.lastLocation,
+                                { _, rawLocation ->
+                                    notificationChannel.offer(NavigationTripDescriptor(NAVIGATION_SERVICE_ID,
+                                            status.routeProgress, rawLocation, status.enhancedLocation))
+                                })
+                    }
+
+                    override fun onFailure(exception: Exception) {
+                    }
+                })
+                delay(UPDATE_INTERVAL_IN_MILLISECONDS) // Poll every second
+            }
+        }
+        locationEngine?.requestLocationUpdates(locationEngineRequest, locationEngineCallback, null)
+        super.onCreate()
+    }
+
+    override fun onDestroy() {
+        stopForeground(true)
+        notificationChannel.close()
+        super.onDestroy()
+    }
+
+    private var notificationChannel = ConflatedBroadcastChannel<NavigationTripDescriptor>()
+    fun getNotificationChannel() = notificationChannel.openSubscription()
+}
+
+/**
+ *TripServiceImpl(val callback:() ->Unit, private val context: Context)
+ * The callback() must contain at least the following code:
+
+val intent: Intent = Intent(context, NavigationNotificationService::class.java)
+try {
+context.startService(intent)
+} catch (e: IllegalStateException) {
+if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+context.startForegroundService(intent)
+} else {
+throw e
+}
+}
+private val navigator: MapboxNativeNavigator,
+
+ */
+internal class TripServiceImpl(val callback: () -> Unit) : TripService {
```

Comment:
Could you clarify how this callback is going to be used? It seems that (in practice for the Android implementation) is going to actually start the service (based on the commented out code above 👀 https://github.com/mapbox/mapbox-navigation-android/pull/2320/files#r358764141) and should include `startService` so that the service is properly created. Is that correct?

Reply:
Correct. The callback contains code required to start the service. Didn't want to make an explicit reference to the service class. Also wanted to allow the caller to do other things in addition to starting the app.